### DISCUSSION TITLE
Sending websocket message from http route

### DISCUSSION THREAD

**Discussion by MateuszBelczowski** [fastapi/fastapi#7986] [Questions] [answered]
**Description**
Let's say that I want to notify client application about the changes performed by other users. 
To do that, I would like to create a websocket endpoint (e.g. /notify) that user can connect to and listen for incoming updates from server.

However, I wasn't able to find any example of doing that in the documentation. What I'm looking for is an easy way to send websocket message to the user from HTTP route. Do you have any suggestions, how can I achieve that?

**Comment by CINOAdam** [comment]
Hi there, I was able to get something like this working using Redis with Socket IO with Fast API.

https://python-socketio.readthedocs.io/en/latest/server.html#emitting-from-external-processes

**Comment by MateuszBelczowski** [comment]
Yes, that's what I meant, thanks :) Any other packages that can be used to achieve this?

**Comment by fcarron** [comment, marked as answer]
Hello

I think that you could keep track of all websocket clients (like it is done in the websockets package (not the starlette one) [https://websockets.readthedocs.io/en/stable/intro.html](url).

Doing so would look like this:

```
import asyncio
from fastapi import FastAPI, Depends
from starlette.websockets import WebSocket

ws_clients = []
app = FastAPI()


def get_ws_clients():
    return ws_clients


@app.get("/update")
async def notify(clients=Depends(get_ws_clients)):
    """
    Notify all websocket clients
    """
    await asyncio.wait([ws.send_text("notification!") for ws in clients])
    return {}


@app.websocket("/notify")
async def websocket_endpoint(websocket: WebSocket, clients=Depends(get_ws_clients)):
    await websocket.accept()
    clients.append(websocket)
    try:
        while True:
            _ = await websocket.receive_text()
    finally:
        ws_clients.remove(websocket)

```
Note that this will only work if the app is running in a single process . Perhaps you should also perform some error handling when sending the notification. (?)

**Reply by YuriiMotov** [reply]
This simplified code example will work if you have only 1 worker. If you need more general solution, use distributed pub\sub like [Redis](https://redis.io/docs/latest/develop/pubsub/), [RabbitMQ](https://www.rabbitmq.com/tutorials/tutorial-three-python), Kafka, etc.. to send messages from `notify` endpoint to `websocket_endpoint`.

**Comment by tiangolo** [comment]
Thanks everyone for the help here!

@MateuszBelczowski does that solve your problem? Do you want to close the issue?

**Comment by MateuszBelczowski** [comment]
@tiangolo Yes, I was used to Django Channels to handle that and was curious what are the alternatives in the FastAPI world. So thanks everyone for the feedback :)

**Comment by Gaston-Ortiz** [comment]
Hello,

I am using the same example that @fcarron recommended. Now I would like to create some unit testing for the API "/update" and I doing that with this code:

`
@app.get("/update")
async def notify(clients=Depends(get_ws_clients)):
    """
    Notify all websocket clients
    """
    await asyncio.wait([ws.send_text("notification!") for ws in clients])
    return {'test': 'test1'}

from fastapi.testclient import TestClient

client = TestClient(app)

def test_update():
    # create the websocket
    with client.websocket_connect("/notify") as ws:
        # Do the request to update
        response = client.get("/update")
        #get the response
        result_dict = response.json()
        print("RESPONSE JSON:", result_dict)
        # validate the outputs
        assert response.status_code == 200
        assert result_dict == {'test': 'test1'}
        # close the websocket conection
        ws.close()
`


        
**But when I run pytest to check the assert I get the error WebSocketDisconnect. This is the output in  the terminal:**



`test_main.py::test_update FAILED                                                                                 [100%]

====================================================== FAILURES =======================================================
_____________________________________________________ test_update _____________________________________________________

    def test_update():
        # create the websocket
        with client.websocket_connect("/notify") as ws:
            # Do the request to update
            response = client.get("/update")
            #get the response
            result_dict = response.json()
            print("RESPONSE JSON:", result_dict)
            # validate the outputs
            assert response.status_code == 200
            assert result_dict == {'test': 'test1'}
            # close the websocket conection
>           ws.close()

test_main.py:50:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\Users\gasto\miniconda3\envs\ai_ml\lib\site-packages\starlette\testclient.py:326: in __exit__
    raise message
C:\Users\gasto\miniconda3\envs\ai_ml\lib\site-packages\anyio\from_thread.py:187: in _call_func
    retval = await retval
C:\Users\gasto\miniconda3\envs\ai_ml\lib\site-packages\starlette\testclient.py:336: in _run
    await self.app(scope, receive, send)
C:\Users\gasto\miniconda3\envs\ai_ml\lib\site-packages\fastapi\applications.py:259: in __call__
    await super().__call__(scope, receive, send)
C:\Users\gasto\miniconda3\envs\ai_ml\lib\site-packages\starlette\applications.py:112: in __call__
    await self.middleware_stack(scope, receive, send)
C:\Users\gasto\miniconda3\envs\ai_ml\lib\site-packages\starlette\middleware\errors.py:146: in __call__
    await self.app(scope, receive, send)
C:\Users\gasto\miniconda3\envs\ai_ml\lib\site-packages\starlette\exceptions.py:58: in __call__
    await self.app(scope, receive, send)
C:\Users\gasto\miniconda3\envs\ai_ml\lib\site-packages\fastapi\middleware\asyncexitstack.py:21: in __call__
    raise e
C:\Users\gasto\miniconda3\envs\ai_ml\lib\site-packages\fastapi\middleware\asyncexitstack.py:18: in __call__
    await self.app(scope, receive, send)
C:\Users\gasto\miniconda3\envs\ai_ml\lib\site-packages\starlette\routing.py:656: in __call__
    await route.handle(scope, receive, send)
C:\Users\gasto\miniconda3\envs\ai_ml\lib\site-packages\starlette\routing.py:315: in handle
    await self.app(scope, receive, send)
C:\Users\gasto\miniconda3\envs\ai_ml\lib\site-packages\starlette\routing.py:77: in app
    await func(session)
C:\Users\gasto\miniconda3\envs\ai_ml\lib\site-packages\fastapi\routing.py:274: in app
    await dependant.call(**values)
test_main.py:28: in websocket_endpoint
    _ = await websocket.receive_text()
C:\Users\gasto\miniconda3\envs\ai_ml\lib\site-packages\starlette\websockets.py:85: in receive_text
    self._raise_on_disconnect(message)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <starlette.websockets.WebSocket object at 0x0000021196B095E0>
message = {'code': 1000, 'type': 'websocket.disconnect'}

    def _raise_on_disconnect(self, message: Message) -> None:
        if message["type"] == "websocket.disconnect":
>           raise WebSocketDisconnect(message["code"])
E           starlette.websockets.WebSocketDisconnect: 1000

C:\Users\gasto\miniconda3\envs\ai_ml\lib\site-packages\starlette\websockets.py:80: WebSocketDisconnect
------------------------------------------------ Captured stdout call -------------------------------------------------
RESPONSE JSON: {'test': 'test1'}
=============================================== short test summary info ===============================================
FAILED test_main.py::test_update - starlette.websockets.WebSocketDisconnect: 1000
================================================== 1 failed in 2.11s ==================================================`




**I tried different approaches but always get the same error. Please can you help me with that? I will really appreciate it.**

**Comment by xiki808** [comment]
This works for me when running on uvicorn, but not when using gunicorn. The list of clients keep resetting

Edit:
I figure it's due to multiple workers. Is there a way how to solve this?

**Comment by s-weigand** [comment]
If you need support for multiple workers you could use [broadcaster](https://github.com/encode/broadcaster) as [the docs recommend in the Tip](https://fastapi.tiangolo.com/advanced/websockets/#handling-disconnections-and-multiple-clients)

**Comment by JarroVGIT** [comment]
Broadcaster does not support multiple workers per se. It holds a central object in memory for all connections, but when you have multiple workers, the connections are scattered across workers. The principles are pretty much the same for the example code provided in the docs.

**Comment by s-weigand** [comment]
@JarroVGIT While I understand that the memory backend would have the same problem since each worker creates its own `async.Queue` which isn't shared with other workers. It should work (not tested) with other backends (e.g. redis) where each worker would subscribe to the same queue 🤔

**Comment by JarroVGIT** [comment]
Well, that is very true. My point wasn't very explicit I realise; it was that Broadcaster doesn't 'solve' the issue of multiple workers. Let's say you have an endpoint to which someone can post a message that need to be broadcasted to all WS clients. This cannot be done in multiple workers, as the receiving endpoint (the one that gets the post with the message to broadcast) only lives in 1 worker. 

However, for external queues, Broadcast can definitely work well. But, that wouldn't really solve the issue of @xiki808 I would think. But then again, I have no idea how having multiple workers would lead to a suddenly empty list (other then unwanted restarts of the workers) :P

**Comment by xiki808** [comment]
Thanks @JarroVGIT, exactly what I would like to solve. My only way to fix this is to resort to one worker, for just one functionality that needs to send a message to all clients connected to the backend.

**Comment by JarroVGIT** [comment]
Well, there might be a solution. Just to confirm I have the problem right: you can only have one worker that listens to something and broadcasts to all connected clients, correct? The problem lies in that you can only have one active worker listening, not in that clients are connecting to different workers. 

If that is the case, you could go for a multi-layered approach in which you have two sub applications:

1. `Application A` is running on its own uvicorn/gunicorn instance with only 1 worker. It is a simple "listen and broadcast" application such as you have now. This application is not public and only serves to broadcast to instances of Application B.
2. `Application B` is also running on its own uvicorn/gunicorn instance but with multiple workers. External clients connect **only** to application B. Application B connects to the websocket endpoint of Application A and for each incoming message, distributes that to all connected clients. 

This way, you are much more scalable while still dealing with the limitation that only 1 worker can serve as a listener to the source.

Anyway, this is way beyond the scope of this issue. I hope it helps but if you run into any problems, please open up a new one.

**Comment by xiki808** [comment]
That's very interesting, thank you! I can try out the multi-layered approach you mentioned. 

One thing I would like to clarify. From my understanding, is that all active workers will be actually listening, but clients are connecting to random workers.

On connection I'm appending the client's websocket object into a List. Later I learnt that the List is not the same in all the workers. So when at some point a client sends a request to the backend, which runs a process where it sends a websocket message to all clients, it is indeed only sending to those clients that have connected to the same worker.

I can see now how I can solve this with what you have suggested, Thank you!

**Comment by xiki808** [comment]
@JarroVGIT quite a while ago now, but finally I found some time to work on the solution based on how you explained it. I'm sorry to open up this again but I'm very intrigued. 

I am scratching my head on how I should send and store the websocket client in `Application A`. 

One thing I have tried is to grab the websocket client object once a client connects to `Application B` and send it over to `Application A`, to which later I figured that I cannot do so, as the websocket object cannot be converted into json.

Another is to redirect the websocket to `Application A` but of course later I understood that I'm actually storing different websocket client object within this application, so I cannot return to it to remove it or send a message directly to the client.

Would you have any ideas how to better solve this?

**Comment by JarroVGIT** [comment]
I think I either got your problem wrong, or you didn't quite get the solution right. I didn't mean to imply to transfer or forward any objects from `Application A` to `Application B`, just to send messages. If you can only have 1 listener (because of whatever limitation you have), you make `Application A` listen to that source. That source emits messages, and `Application A` catches those messages. `Application A` also provides a websocket endpoint, to which it emits the caught messages. 

`Application B` has a similar setup, but rather than listening to the source, it will listen to the websocket endpoint that is provided by `Application A`. `Application B` also offers a websocket endpoint, and any messages caught from the websocket endpoint of Application A, is broadcasted to all the connected clients that connected to the websocket endpoint of `Application B`. 

Note, this setup has only one purpose: to be able to scale (reasonably) the amount of possible end users that can listen to a source that can only handle **one** listener. So, any message from a **source** will be emitted to `Application A` (one instance only), which will broadcast it to all it's clients (which are `Application B` instances!), and each `Application B` instance that is listening to `Application A` will broadcast the message to its own (end users) connected clients.

Hope this clarifies :)

**Comment by xiki808** [comment]
I know what you mean now. Thank you for following up and giving me a proper feedback. 

I was under the impression that `Application A` is not publicly available, so the source has to connect to `Application B` to get to `A` in which it keeps memory of all the clients under one worker.
